bundles/org.openhab.binding.elroconnects/src/main/java/org/openhab/binding/elroconnects/internal/handler/ElroConnectsBridgeHandler.java

```diff
@@ -0,0 +1,880 @@
+        socket.send(queryPacket);
+    }
+
+    private String receive(DatagramSocket socket, byte[] buffer, DatagramPacket packet) throws IOException {
+        socket.receive(packet);
+        String response = new String(packet.getData(), packet.getOffset(), packet.getLength());
+        logger.debug("Received: {}", response);
+        addr = packet.getAddress();
+        return response;
+    }
+
+    /**
+     * Basic method to send an {@link ElroConnectsMessage} on an open socket with the K1 hub.
+     *
+     * @param socket
+     * @param elroMessage
+     * @param waitResponse true if no new messages should be allowed to be sent before receiving the full response
+     * @throws IOException
+     */
+    private synchronized void sendElroMessage(DatagramSocket socket, ElroConnectsMessage elroMessage,
+            boolean waitResponse) throws IOException {
+        String message = gsonOut.toJson(elroMessage);
+        awaitResponse(waitResponse);
+        send(socket, message, false);
+    }
+
+    /**
+     * Send device control command. The device command string various by device type. The calling method is responsible
+     * for creating the appropriate command string.
+     *
+     * @param deviceId
+     * @param deviceCommand Elro Connects device command string
+     * @throws IOException
+     */
+    public void deviceControl(int deviceId, String deviceCommand) throws IOException {
+        String connectorId = this.connectorId;
+        String ctrlKey = this.ctrlKey;
+        DatagramSocket socket = this.socket;
+        if (!connectorId.isEmpty() && !ctrlKey.isEmpty() && (socket != null)) {
+            logger.debug("Device control {}, status {}", deviceId, deviceCommand);
+            ElroConnectsMessage elroMessage = new ElroConnectsMessage(msgIdIncrement(), connectorId, ctrlKey,
+                    ELRO_DEVICE_CONTROL).withDeviceId(ElroConnectsUtil.encode(deviceId))
+                            .withDeviceStatus(deviceCommand);
+            sendElroMessage(socket, elroMessage, false);
+        } else {
+            // This should not happen if properly initialized
+            logger.debug("Could not send");
+        }
+    }
+
+    /**
+     * Send request to receive all device names.
+     *
+     * @throws IOException
+     */
+    private void getDeviceNames() throws IOException {
+        String connectorId = this.connectorId;
+        String ctrlKey = this.ctrlKey;
+        DatagramSocket socket = this.socket;
+        if (!connectorId.isEmpty() && !ctrlKey.isEmpty() && (socket != null)) {
+            logger.debug("Get device names");
+            ElroConnectsMessage elroMessage = new ElroConnectsMessage(msgIdIncrement(), connectorId, ctrlKey,
+                    ELRO_GET_DEVICE_NAME).withDeviceId(0);
+            sendElroMessage(socket, elroMessage, true);
+        } else {
+            // This should not happen if properly initialized
```

Comment:
Should the thing go OFFLINE with a config error message? Another way is to make the config fields REQUIRED so that the handler is not created by the framework until the user enters in values.